Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp

```diff
@@ -0,0 +1,153 @@
+
+    ~_WebKitQueryPermissionRequest()
+    {
+        // Fallback to Prompt response unless the completion handler was already called.
+        if (completionHandler)
+            completionHandler(WebCore::PermissionState::Prompt);
+    }
+
+    CString permissionName;
+    CompletionHandler<void(std::optional<WebCore::PermissionState>)> completionHandler;
+    RefPtr<API::SecurityOrigin> securityOrigin;
+    int referenceCount { 1 };
+};
+
+G_DEFINE_BOXED_TYPE(WebKitQueryPermissionRequest, webkit_query_permission_request, webkit_query_permission_request_ref, webkit_query_permission_request_unref)
+
+WebKitQueryPermissionRequest* webkitQueryPermissionRequestCreate(const WTF::String& permissionName, API::SecurityOrigin& origin, CompletionHandler<void(std::optional<WebCore::PermissionState>)>&& completionHandler)
+{
+    WebKitQueryPermissionRequest* queryPermissionRequest = static_cast<WebKitQueryPermissionRequest*>(fastMalloc(sizeof(WebKitQueryPermissionRequest)));
+    new (queryPermissionRequest) WebKitQueryPermissionRequest(permissionName, origin, WTFMove(completionHandler));
+    return queryPermissionRequest;
+}
+
+WebKitQueryPermissionRequest*
+webkit_query_permission_request_ref(WebKitQueryPermissionRequest* queryPermissionRequest)
+{
+    g_return_val_if_fail(queryPermissionRequest, nullptr);
+
+    g_atomic_int_inc(&queryPermissionRequest->referenceCount);
+    return queryPermissionRequest;
+}
+
+void webkit_query_permission_request_unref(WebKitQueryPermissionRequest* queryPermissionRequest)
+{
+    g_return_if_fail(queryPermissionRequest);
+
+    if (g_atomic_int_dec_and_test(&queryPermissionRequest->referenceCount)) {
+        queryPermissionRequest->~WebKitQueryPermissionRequest();
+        fastFree(queryPermissionRequest);
+    }
+}
+
+/**
+ * webkit_query_permission_get_name:
+ * @request: a #WebKitQueryPermissionRequest
+ *
+ * Get the permission name for which access is being requested.
+ *
+ * Returns: the permission name for @request
+ *
+ * Since: 2.40
+ */
+const gchar*
+webkit_query_permission_get_name(WebKitQueryPermissionRequest* request)
+{
+    g_return_val_if_fail(request, nullptr);
+
+    return request->permissionName.data();
+}
+
+/**
+ * webkit_query_permission_get_security_origin:
+ * @request: a #WebKitQueryPermissionRequest
+ *
+ * Get the permission origin for which access is being requested.
+ *
+ * Returns: (transfer full): A new #WebKitSecurityOrigin representing the origin from which the
+ * @request was emitted.
+ *
+ * Since: 2.40
+ */
+WebKitSecurityOrigin *
+webkit_query_permission_get_security_origin(WebKitQueryPermissionRequest* request)
+{
+    g_return_val_if_fail(request, nullptr);
+
+    return webkitSecurityOriginCreate(request->securityOrigin->securityOrigin().isolatedCopy());
+}
+
+/**
+ * webkit_permission_query_finish:
+ * @request: a #WebKitQueryPermissionRequest
+ * @state: a #WebKitPermissionState
+ *
+ * Notify the web-engine of the selected permission state for the given request.
```

Comment:
Add here that this should only be called as a response to the `WebKitWebView::query-permission` signal